### ISSUE TITLE
Subnets Not being calculated correctly

### ISSUE THREAD

**Issue by mkriedel** [pulumi/pulumi-awsx#513]
Given the following `VpcArgs` when constructing an AWSX VPC, I'm experiencing an issue where the subnet topology isn't being computed correctly:
```javascript
{
            cidrBlock: '10.0.0.0/20',
            subnets: [
                {
                    type: "public",
                    cidrMask: 24
                },
                {
                    type: "private",
                    cidrMask: 22
                }
            ],
            numberOfAvailabilityZones: 3,
}
```
What happens:
I get 3 public and 3 private subnets:
* `10.0.0.0/24` - public
* `10.0.1.0/24` - public
* `10.0.2.0/24` - public
* `10.0.3.0/22` - private
* `10.0.7.0/22` - private
* `10.0.11.0/22` - private

This means that we're creating overlapping subnets (`10.0.3.0/22` overlaps with the first 3 public subnets since that represents a subnet that goes from `10.0.0.0 - 10.0.3.255`).

What I expect to happen:
I expect that specifying `cidrMask` will result in a valid subnet topology, such as:
* `10.0.0.0/24` - public
* `10.0.1.0/24` - public
* `10.0.2.0/24` - public
* `10.0.3.0/24` - unallocated (just don't create it)
* `10.0.4.0/22` - private
* `10.0.8.0/22` - private
* `10.0.12.0/22` - private

There should at least be a validation that the computed topology doesn't contain overlapping subnets.

Workaround:
I created the larger subnets first, which created clean alignments with subnet boundaries.

**Comment by stack72** [comment]
@jkodroff i am pretty confident that the work you did in the new VPC rewrite takes care of things and ensures we don’t have a relapsing subnets, right?

**Comment by zscott** [comment]
Is there an update on this? A workaround would suffice. I am running into this as well.

Unfortunately, in my case, all of the subnets are the same size.

```
    const subnetSpecs: awsx.types.input.ec2.SubnetSpecArgs[] = [
        {
            type: awsx.ec2.SubnetType.Public,
            name: "public",
            cidrMask: 24,
            tags: {
                ...publicSubnetTags,
            },
        },
        {
            type: awsx.ec2.SubnetType.Private,
            name: "kubernetes",
            cidrMask: 24,
            tags: {
                ...privateSubnetTags,
            },
        },
        {
            type: awsx.ec2.SubnetType.Private,
            name: "rds",
            cidrMask: 24,
            tags: rdsSubnetTags,
        },
    ];
    ```
    
    Thank you.

**Comment by zscott** [comment]
Just learned that the problem only happens when I initially create the VPC with 2 of the 3 subnets, then update it by adding the 3rd.

When I destroy the vpc and run pulumi up from scratch it works.

**Comment by jkodroff** [comment]
@zscott Sorry this caused you trouble. Just to set expectations, it's probably going to be very difficult or impossible to get CIDR blocks to recalculate on an update without causing collisions/needing replacement.

**Comment by zscott** [comment]
Thanks.

Might be cool if there was a way I could inject a typescript class where we
could provide our own custom logic for this. In my head I know how I’d do
it but agree that it’s hard or impossible to make a one size fits all
solution.

**Comment by thomas11** [comment]
It seems like the original bug was fixed by #792 and #814, and the additional question of adding subnets later would be hard to implement. I'll close this issue, then - feel free to re-open if this is in error.
